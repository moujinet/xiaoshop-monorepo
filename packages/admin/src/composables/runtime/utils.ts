import type { RouteRecordRaw } from 'vue-router/auto'
import type { IMenu } from '~/types'

/**
 * Transform id to path
 *
 * @param id string
 * @returns string
 */
export function transId2Path(id: string): string {
  return `/${id.split('.').splice(1).join('/')}`
}

/**
 * Extend routes meta
 *
 * @param userRoutes RouteRecordRaw[]
 * @param menus IMenu[]
 *
 * @returns RouteRecordRaw[]
 */
export function extendRoutesMeta(userRoutes: RouteRecordRaw[], menus: IMenu[]): RouteRecordRaw[] {
  const routes: RouteRecordRaw[] = []

  userRoutes.forEach((r) => {
    const route = { ...r }
    const meta = _deepFind(menus, route)

    if (meta) {
      route.meta = {
        ...route.meta,
        id: meta.id,
        space: meta.space,
        module: meta.module,
        title: meta.name,
        desc: meta.desc,
        icon: meta.icon,
        isPermission: meta.isPermission,
      }
    }

    if (route.children && route.children.length > 0)
      route.children = extendRoutesMeta(route.children, menus) || []

    routes.push(route)
  })

  return routes
}

/**
 * Deep find menu by route name
 *
 * @param menus IMenu[]
 * @param route RouteRecordRaw
 *
 * @returns IMenu | undefined
 */
function _deepFind(menus: IMenu[], route: RouteRecordRaw): IMenu | undefined {
  const target = ((route.name || route.path) as string).replace(/\/+$/, '')

  let matched: IMenu | undefined

  for (const menu of menus) {
    if (transId2Path(menu.id) === target) {
      matched = menu
      break
    }

    if (menu.children && menu.children.length > 0)
      matched = _deepFind(menu.children, route) || matched
  }

  return matched
}
