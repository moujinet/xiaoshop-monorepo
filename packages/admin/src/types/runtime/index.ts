export * from './utils'
export * from './middleware'
export * from './plugin'
export * from './global'
export * from './space'
export * from './module'
export * from './menu'
export * from './meta'
