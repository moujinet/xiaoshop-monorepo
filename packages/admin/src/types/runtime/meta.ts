export interface IMeta {
  id: string
  space: string
  module: string
  name: string
  desc: string
  icon: string
}
