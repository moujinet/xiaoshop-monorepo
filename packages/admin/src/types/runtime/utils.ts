import type { IGlobalContext } from './global'

export interface ICallback<T = void> {
  (): T
}

export type IAsyncCallback<T = void> = Promise<T>

export interface IContextCallback<T = void> {
  (ctx: IGlobalContext): T
}
