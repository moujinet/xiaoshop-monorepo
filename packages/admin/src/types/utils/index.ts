export interface IMockSetup {
  (): void
}

export interface IMockRequestOptions {
  method: string
  url: string
  body: any
  query: any
}

export interface IMockDefinition {
  [url: string]: (options: IMockRequestOptions) => any
}
