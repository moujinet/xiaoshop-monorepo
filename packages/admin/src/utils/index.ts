import dayjs from 'dayjs'
import { IS_DEBUG_MODE } from '~/constants/env'

/**
 * 问候语
 *
 * @returns string
 */
export function getGreeting() {
  const hour = new Date().getHours()

  return hour < 9
    ? '早上好'
    : hour <= 11
      ? '上午好'
      : hour <= 13
        ? '中午好'
        : hour < 20
          ? '下午好'
          : '晚上好'
}

/**
 * 返回指定格式日期
 *
 * @param date Dayjs | number | string | undefined
 * @param format string
 * @returns string
 */
export function formatDateTime(
  date: dayjs.Dayjs | number | string | undefined = undefined,
  format = 'YYYY-MM-DD HH:mm:ss',
): string {
  return dayjs(date).format(format)
}

/**
 * 输出调试信息
 *
 * @param type
 * @param message
 * @param context
 */
export function debug(type: string, message: string, context: any = null) {
  IS_DEBUG_MODE
  // eslint-disable-next-line no-console
  && console.log(
    '[XIAOSHOP]',
    type.toUpperCase(),
    message,
    context,
    formatDateTime(new Date().getTime(), 'YY-MM-DD hh:mm:ss.SSS'),
  )
}

/**
 * 对象去除某些属性
 *
 * @param obj T
 * @param keys K
 * @returns Omit<T, K | K2>
 */
export function omit<
  T extends Record<string, any>,
  K extends string,
  K2 extends keyof T,
>(obj: T, keys: (K | K2)[]) {
  const result = { ...obj }

  keys.forEach((key) => {
    delete result[key]
  })

  return result as Omit<T, K2>
}

/**
 * 对象选取某些属性
 *
 * @param obj
 * @param keys
 * @returns Pick<T, K | K2>
 */
export function pick<
T extends Record<string, any>,
K extends string,
K2 extends keyof T,
>(obj: T, keys: (K | K2)[]): Pick<T, K2> {
  return keys.reduce((pre, cur) => {
    if (cur in obj)
      pre[cur] = obj[cur]
    return pre
  }, {} as Pick<T, K | K2>)
}
