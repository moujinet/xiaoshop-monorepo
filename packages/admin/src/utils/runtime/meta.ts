import type { RouteRecordRaw } from 'vue-router/auto'
import type { IMenu, IMeta } from '~/types'

export const useMeta = defineStore('meta', () => {
  /**
   * 元信息，用于扩展路由 (内部)
   * 将空间、模块及菜单的基础信息放到元信息中
   */
  const _metas = ref<IMeta[]>([])

  /**
   * 返回已排序的元信息
   */
  const metas = computed(() => {
    if (_metas.value.length === 0)
      refreshMetas()

    return _metas.value.sort((a, b) => a.id.localeCompare(b.id))
  })

  /**
   * 返回指定元信息
   *
   * @param id string
   * @returns IMeta | undefined
   */
  function getMeta(id: string): IMeta | undefined {
    return metas.value.find(m => m.id === id)
  }

  /**
   * 扩展路由 Meta
   *
   * @param userRoutes RouteRecordRaw[]
   * @returns RouteRecordRaw[]
   */
  function extendRoutesMeta(userRoutes: RouteRecordRaw[], parentPath = ''): RouteRecordRaw[] {
    const routes: RouteRecordRaw[] = []
    const _parentPath = parentPath

    userRoutes.forEach((r) => {
      const route = { ...r }
      const meta = metas.value.find((m) => {
        if (route.path === '' || route.path === '/')
          return false

        if (
          route.path.startsWith('/')
          || (route.name && (route.name as string).startsWith('/'))
        ) {
          return transId2Path(m.id) === route.name
            || `${transId2Path(m.id)}/` === route.name
            || transId2Path(m.id) === route.path
        }

        return transId2Path(m.id) === `${parentPath}/${route.path}`
      })

      if (meta) {
        parentPath += route.path.startsWith('/') ? route.path : `/${route.path}`

        route.meta = {
          ...route.meta,
          id: meta.id,
          space: meta.space,
          module: meta.module,
          name: meta.name,
          desc: meta.desc,
          icon: meta.icon,
        }
      }

      if (route.children && route.children.length > 0)
        route.children = extendRoutesMeta(route.children, parentPath)

      parentPath = meta ? _parentPath : ''

      routes.push(route)
    })

    return routes
  }

  /**
   * 刷新元信息
   */
  function refreshMetas() {
    // Clear all
    _metas.value = []

    // Spaces
    _refreshSpaces()
    // Modules
    _refreshModules()
    // Menus
    _refreshMenus()
  }

  /**
   * 刷新空间元信息
   */
  function _refreshSpaces() {
    const { spaces } = storeToRefs(useSpaces())

    spaces.value.forEach((space) => {
      _metas.value.push({
        id: space.id,
        space: '',
        module: '',
        name: space.name,
        desc: space.desc || '',
        icon: space.icon,
      })
    })
  }

  /**
   * 刷新模块元信息
   */
  function _refreshModules() {
    const { modules } = storeToRefs(useModules())

    modules.value.forEach((module) => {
      _metas.value.push({
        id: module.id,
        space: module.space,
        module: '',
        name: module.name,
        desc: module.desc || '',
        icon: module.icon,
      })
    })
  }

  /**
   * 刷新菜单元信息
   */
  function _refreshMenus() {
    const { menus } = storeToRefs(useMenus())
    _refreshMenusChildren(menus.value)
  }

  /**
   * 刷新菜单子元信息
   *
   * @param children IMenu[]
   */
  function _refreshMenusChildren(children: IMenu[]) {
    children.forEach((menu) => {
      _metas.value.push({
        id: menu.id,
        space: menu.space,
        module: menu.module,
        name: menu.name,
        desc: menu.desc || '',
        icon: menu.icon,
      })

      if (menu.children && menu.children.length > 0)
        _refreshMenusChildren(menu.children)
    })
  }

  return {
    /**
     * 所有元信息
     */
    metas,
    /**
     * 返回指定元信息
     */
    getMeta,
    /**
     * 刷新元信息
     */
    refreshMetas,
    /**
     * 扩展路由 Meta
     */
    extendRoutesMeta,
  }
})
