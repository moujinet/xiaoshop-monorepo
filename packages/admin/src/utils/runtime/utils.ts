/**
 * 返回标准化菜单 ID
 *
 * @param id string
 * @returns string
 */
export function normalizeMenuId(id: string) {
  return id
    .replace(/#/, '')
    .replace(/\.index$/, '')
}

/**
 * 转换菜单 ID 为访问路径
 *
 * @param id string
 * @returns string
 */
export function transId2Path(id: string): string {
  const parts: string[] = id.split('.')

  if (parts[parts.length - 1] === 'index')
    parts.pop()

  return `/${parts.splice(1).join('/')}`
}
