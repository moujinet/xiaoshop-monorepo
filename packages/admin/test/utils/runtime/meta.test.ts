import type { RouteRecordRaw } from 'vue-router/auto'
import { describe, expect, it } from 'vitest'
import { goodsRoutes, metas } from './_fixtures/meta'
import { transId2Path } from '~/utils/runtime/utils'

function extendRoutesMeta(userRoutes: RouteRecordRaw[], parentPath = ''): RouteRecordRaw[] {
  const routes: RouteRecordRaw[] = []
  const _parentPath = parentPath

  userRoutes.forEach((r) => {
    const route = { ...r }
    const meta = metas.find((m) => {
      if (route.path === '' || route.path === '/')
        return false

      if (route.path.startsWith('/') || (route.name && (route.name as string).startsWith('/'))) {
        return transId2Path(m.id) === route.name
          || `${transId2Path(m.id)}/` === route.name
          || transId2Path(m.id) === route.path
      }

      return transId2Path(m.id) === `${parentPath}/${route.path}`
    })

    if (meta) {
      parentPath += route.path.startsWith('/') ? route.path : `/${route.path}`

      route.meta = {
        ...route.meta,
        id: meta.id,
        space: meta.space,
        module: meta.module,
        name: meta.name,
        desc: meta.desc,
        icon: meta.icon,
      }
    }

    if (route.children && route.children.length > 0)
      route.children = extendRoutesMeta(route.children, parentPath)

    parentPath = meta ? _parentPath : ''

    routes.push(route)
  })

  return routes
}

describe('meta', () => {
  it('extendRoutesMeta', () => {
    expect(extendRoutesMeta(goodsRoutes as RouteRecordRaw[])).toMatchSnapshot()
  })
})
