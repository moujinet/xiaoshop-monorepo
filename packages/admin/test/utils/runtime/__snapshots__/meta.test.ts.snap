// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`meta > extendRoutesMeta 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "name": "/goods/manage/attributes/",
                    "path": "",
                  },
                  {
                    "meta": {
                      "desc": "",
                      "icon": "",
                      "id": "shop.goods.manage.attributes.template",
                      "module": "shop.goods",
                      "name": "模板设置",
                      "space": "shop",
                    },
                    "name": "/goods/manage/attributes/template",
                    "path": "template",
                  },
                ],
                "meta": {
                  "desc": "",
                  "icon": "",
                  "id": "shop.goods.manage.attributes",
                  "module": "shop.goods",
                  "name": "商品参数",
                  "space": "shop",
                },
                "path": "attributes",
              },
            ],
            "meta": {
              "desc": "",
              "icon": "ph:package",
              "id": "shop.goods.manage",
              "module": "shop.goods",
              "name": "商品管理",
              "space": "shop",
            },
            "path": "manage",
          },
        ],
        "path": "",
      },
    ],
    "meta": {
      "desc": "商品管理",
      "icon": "ph:package",
      "id": "shop.goods",
      "isLayout": true,
      "module": "",
      "name": "商品",
      "space": "shop",
    },
    "path": "/goods",
  },
]
`;
